File: android/app/src/main/java/com/example/paperclipper/FeedbackClient.kt
package com.example.paperclipper

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL

/** Posts user feedback to the same backend (behind the Cloudflare tunnel) as /analyze. */
object FeedbackClient {
    suspend fun send(message: String, email: String?): Boolean = withContext(Dispatchers.IO) {
        val base = BuildConfig.SERVER_URL.trimEnd('/')
        if (base.isBlank()) return@withContext false
        val body = JSONObject()
            .put("message", message)
            .put("email", email ?: JSONObject.NULL)
            .put("appVersion", BuildConfig.VERSION_NAME)
            .toString()

        var conn: HttpURLConnection? = null
        try {
            conn = (URL("$base/feedback").openConnection() as HttpURLConnection).apply {
                requestMethod = "POST"
                connectTimeout = 20_000
                readTimeout = 30_000
                doOutput = true
                setRequestProperty("Content-Type", "application/json")
                setRequestProperty("Authorization", "Bearer ${BuildConfig.PROXY_TOKEN}")
            }
            conn.outputStream.use { it.write(body.toByteArray(Charsets.UTF_8)) }
            conn.responseCode in 200..299
        } catch (e: Exception) {
            false
        } finally {
            conn?.disconnect()
        }
    }
}

File: ios/PaperClipper/Networking/FeedbackClient.swift
import Foundation

/// Posts user feedback to the same backend (behind the Cloudflare tunnel) as `/analyze`. Mirrors
/// Android's `FeedbackClient.kt`: POST `/feedback` with `{message, email, appVersion}` and the proxy
/// bearer token. Returns whether the server accepted it (2xx).
enum FeedbackClient {
    static func send(message: String, email: String?) async -> Bool {
        guard case let .success(base) = Backend.resolveBaseUrl(AppConfig.serverURL),
              let url = URL(string: Backend.buildUrl(base, "feedback")) else {
            return false
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.timeoutInterval = 30
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(AppConfig.proxyToken)", forHTTPHeaderField: "Authorization")

        var body: [String: Any] = [
            "message": message,
            "appVersion": AppConfig.appVersion,
        ]
        body["email"] = email ?? NSNull()
        request.httpBody = try? JSONSerialization.data(withJSONObject: body)

        do {
            let (_, response) = try await URLSession.shared.data(for: request)
            let code = (response as? HTTPURLResponse)?.statusCode ?? 0
            return (200..<300).contains(code)
        } catch {
            return false
        }
    }
}
